[RFC] Vgetorpeek disect

hardenedapple opened the pull request:
I'm in the middle of trying to understand `vgetorpeek()`, and in order to do so I'm refactoring it into a form I can read easier.

I thought I'd open a PR to see if people want the refactoring in the main branch and to take advantage of the multiple tests runs.

I can see that refactoring in such a low-level function may feel too dangerous too include.

I haven't done much as yet -- most of splitting into separate functions hasn't been started yet.

hardenedapple committed:
Use 'exiting' variable to mark while() loop extent

This is used instead of having a condition on the 'c' variable.
We do this so it's clear at what point the while () loop can exit, e.g.
you can search for 'exiting = ' to find all places where it can happen.

Moreover, in the future we may be able to separate concerns between 'c'
as a variable to determine exiting the loop, as a temporary variable,
and as the value to return.

Remove vgetorpeek() unnecessary setting of c = -1

Before it was used simply to signify continuing on in the loop, it's not
needed any more.

'c' is always reset on each iteration around the loop, any value
leftover from the previous loop will be unused.

hardenedapple committed:
Explicit check for got_int in vgetorpeek() while loop

Before it used the check inside the for(;;) loop which I think is less
clear.
Also add explicit setting of `exiting` to `true` or `false` in each
branch of the main `if()` statement for the while loop.

I don't believe this would change the behaviour in any way, as there
isn't any way that the initial for loop can be broken out of and not
gotten back into. But it's nice to be explicit.

hardenedapple committed:
vgetorpeek swap `break` for `continue`

The difference between using `break` and `continue` is that `break`
puts you into the outer while () loop that checks the typeahead_char
and readbuffers before going back into the main for(;;) loop handling
the typeahead buffer.

This change is valid if there is no benefit to checking the stuffbuffers
and typeahead_char in the places where `break` used to be.
I feel comfortable asserting this is the case, as the methods that the
stuffbuffers can be added to are known and very limited.
The only way the stuffbuffer or typeahead_char can be modified are in
user code evaluating an <expr> mapping (there is no other place in the
vgetorpeek() for(;;) loop that could affect them).
The only way that these hypothetical characters in the stuffbuffer would
be used before characters in the typeahead before this change is if
there were an error looking for characters in the typeahead (either an
integer overflow in the typeahead buffer), an error in the evaluation of
an <expr> mapping, or a mapping that is nested too deeply.

During this evaluation both textlock and ex_normal_lock are both set,
which limits the available code paths.
There are three possible places taht

There are three possible places that can change either the
typeahead_char or the stuffbufers while textlock and ex_normal_lock are
both set.

1) do_more_prompt()
    This sets the typeahead_char when the key ':' is pressed at a
    -- More -- prompt.
    When this is pressed, got_int is also set, which means that this is
    the most likely difference to be observed.
    In order to see this, start vim in a short terminal (i.e. not many
    lines), source the vimrc
        function TestMapping()
          let oldkeymap = &keymap
          set keymap=dvorak
          lmap
          return 'j'
          let &keymap = oldkeymap
        endfunction
        nmap <expr> ab TestMapping()
    press `ab`, see the -- More -- prompt, and press ':'.
    Before this commit you'd get the 'Type  :quit<Enter>  to exit Nvim'
    message, and the map isn't used.
    After this commit the map still isn't used, but you get to the ':'
    prompt instead of getting the 'Type  :quit<Enter>  to exit Nvim'
    message.

2) get_number() and get_user_input()
    These put text into the stuffbuffer.
    In order for the difference between using `continue` and `break` to be
    observed, eval0() must return FAIL.
    This requires some error to have happened, as an error must have happened,
    emsg() is called (either in something lower than eval0(), or eval0()
    itself). emsg() calls flush_buffers() (either indirectly or directly),
    unless emsg_off is set.
    When flush_buffers() is set, both the stuffbuffers and typebuf are cleared,
    so this change doesn't affect anything.

hardenedapple committed:
Split uses of 'i' in vgetorpeek()

It was used in two places for two completely separate reasons, with a
previous change one of those reasons became obviously superfluous.

hardenedapple committed:
exiting = (long condition)  is now   exiting = true

The condition looked at whether the integer 'c' was less than zero, or
NUL at the same time that 'advance' was set.
This integer was gotten with 'typebuf.tb_buf[typebuf.tb_off] & 255',
inside a clause that checked typebuf.tb_len and no code between entry
to that clause and this point changed typebuf.
Hence, because typebuf.tb_len was non-zero, the character at
typebuf.tb_buf[typebuf.tb_off] has to be positive (NUL characters are
never inserted into the typebuf).

hardenedapple committed:
Put some code into separate blocks

Just clearly demonstrates that the variables declared inside that block
aren't used anywhere else.
Also move new_wcol and new_wrow into the block where they're used and
distinguish between independent uses of old_w{col,row}

hardenedapple committed:
Remove the while() loop in vgetorpeek()

It's just not needed (because of the changes I've made previously).
Can now use `break` statements in the for(;;) loop instead of `break`
and setting `exiting = true`.

hardenedapple committed:
Move declarations into the blocks they're used

Also block off wait_tb_len into its only use.

justinmk commented:
👍  Try to keep the PR as small as possible, and avoid style changes where a line would otherwise be untouched. Then we can merge (and review more easily) the incremental steps.

hardenedapple commented:
Just a note for anyone viewing the commits: adding `?w=1` to the end of a commit URL gives the diff ignoring whitespace.

This is especially useful for those commits adding/removing a block separation.

hardenedapple committed:
Move checking the typebuf into a separate function

Also this makes it clear 'c' is unaffected when looking in typebuf.

The only time 'c' is set in this function is before exiting the for(;;)
loop, so it's not something to worry about when looking at getting the
key from the user.

hardenedapple committed:
Separate getting input from user into function

hardenedapple committed:
Split some functionality into separate functions

Move checking for toggling paste into separate function.
Put expanding a mapping in the typebuf into its own function.

hardenedapple committed:
Put looking for a mapping into separate function

hardenedapple committed:
Remove unnecessary setting of keylen in look_in_typebuf()

*keylenp was set to 0 before an if/else clause.
In the if branch of that clause, look_in_typebuf() returns a value c >
0, which would cause vgetorpeek() to break out of the for loop, hence
keylen will no longer be used.
In the else branch, *keylenp is set to something different, hence
setting *keylenp before was superfluous.

hardenedapple committed:
Remove superfluous else branch

This branch just set keylen to mp_match_len if mp != NULL.
The previous function called was find_typed_map(), and it's this that
determines the value of mp here.
If this function doesn't return NULL for mp, it is because a map was
found.
In this case keylen was set to mp_match_len in find_typed_map(), and
since mp is not NULL, then check_togglepaste() won't have changed
keylen.

hardenedapple committed:
Make a clear short-circuit in check_togglepaste()

If we have a possible map, then this function does nothing -- it's
clearer if that is a single if statement at the start.

hardenedapple committed:
Remove superfluous nested if statement

The block only did anything if (mp == NULL), but was entered
if (mp == NULL ||  <something>), so remove the <something>.

hardenedapple committed:
mp_match_len is now only used in find_typed_map()

After a lot of small changes it becomes apparent that this value
a) is not used anywhere else, and
b) has no stored state over different loop iterations.

hardenedapple committed:
max_mlen is an unused variable -- remove it

hardenedapple committed:
Explicitly check for mp == NULL before pastetoggle

It's a little superfluous, but it means that no-one has to look in
check_togglepaste() to see that if a mapping has been found it does
nothing.
This makes it clearer when viewing look_in_typebuf() that mappings are
treated preferentially.

hardenedapple committed:
get_key_from_user() was using *cp for two things

Separate the uses into measuring the number of bytes read by inchar()
and returning to vgetorpeek() the next character to use.

The only time that it matters what value vgetorpeek() ends up with for
'c' is when it's breaking out of the for loop. This is only when
get_key_from_user() is returning 1.

hardenedapple committed:
'c' in vgetorpeek() doesn't need to be reset each loop

It's never used as a value in the for(;;) loop -- it's just set to the
return value before signalling that vgetorpeek() has to break out of the
loop and return.

hardenedapple committed:
insert-mode special case for ESC separate function

hardenedapple committed:
Remove a separating block

I had added it to demonstrate that showcmd_len, wait_tb_len, and c1,
were only ever used in that block. After more refactoring it's obvious
because this block is now at the end of a function.

hardenedapple committed:
Rename a variable to show what it does

hardenedapple committed:
get_key_from_user() now returns 'c' if we're to finish

Because this block was the last in the for loop, we don't need to
distinguish between `carry on` and `continue;`.
Hence I can substitute all `return 0;` statements with `return -1;` and
use returning `0` as a valid value to set 'c'.

hardenedapple committed:
Slight tidy-up of an if/else clause

Initially it was in the middle of things, so it needed to specify
continue/break/fall-through etc, but now it's at the end of a function,
and given the use of the function, we can remove most flow-control from
it.
This leaves the if/else clause only updating things as required.

hardenedapple committed:
Move pretty_partial_match into if block

First step to putting this in separate function.

hardenedapple committed:
Move handling showcmd into a separate function

hardenedapple commented:
I made a mistake in 90fed0a77c033f74d92d5d3324cc54f1c57e95af that I fixed in 18697a90ae021d4c6e96d57d480329e6fbe3ca07 .

I now have the code mostly readable, so I'm going to slowly rebase to remove that mistake. That's when I'll make sure each commit passes all automated tests.

Just mentioning that here so that anyone having a look at this PR understands where all the commits go after this point.

hardenedapple committed:
Declare variables in block they're used

Also rename some uses to make it clear they're independent of each
other.
Move declarations of mp, mp2, and mp_match
Split 'n' and 'c1' uses into independant variables
Split use of 's' into distict places
Mark vgetorpeek argument as const

hardenedapple committed:
Split some functionality into separate functions

calculate wait time for inchar() in separate function
Handle -- INSERT -- message in separate function
Move testing for interrupts into a function

hardenedapple commented:
That was an accidental close -- I ran `git-push` with the base commit in the hopes the travis tests would be run on something without any changes.

hardenedapple committed:
Misc reordering

Use local_State in the LANGMAP_ADJUST() condition
Separate a long if clause into many
Set `mode_deleted` and `timedout` to be bools

hardenedapple committed:
clint

hardenedapple commented:
While I didn't keep the entire PR small, each step is really small, so reviewing and merging individual commits should be easy.

bfredl commented:
82 commits is a bit much to review commits individually, maybe logically similar commits can be squashed? Also  small stylistic/clint changes can be squashed together with a preceding related change, no need for a separate commit just for joining two lines, for instance.

bfredl commented on src/nvim/getchar.c:
```diff
@@ -1577,729 +1577,843 @@ vungetc ( /* unget one character (can only be done once!) */
-              if ((*mb_ptr2cells)(ptr + col) > 1)
-                --curwin->w_wcol;
+            vcol += lbr_chartabsize(ptr, ptr + col, (colnr_T)vcol);
+            if (has_mbyte) {
```
you can use that `has_mbyte` is always true in nvim.

bfredl commented on src/nvim/getchar.c:
```diff
@@ -1577,729 +1577,843 @@ vungetc ( /* unget one character (can only be done once!) */
-                --curwin->w_wcol;
+            vcol += lbr_chartabsize(ptr, ptr + col, (colnr_T)vcol);
+            if (has_mbyte) {
+              col += (*mb_ptr2len)(ptr + col);
```
Also with larger refactorings we should change `(*mb_func)(` and `MB_FUNC(` to to `mb_func(`. The only reason against would be merging from vim easier but with heavy refactoring applying patches must be done manually anyway.

hardenedapple committed:
mbyte stuff recommended by bfredl

hardenedapple commented:
I've merged a bunch of logically similar commits together as suggested.

aktau commented on src/nvim/getchar.c:
```diff
@@ -1577,6 +1577,73 @@ vungetc ( /* unget one character (can only be done once!) */
+  return p_tm;
+}
+
+static int handle_int(const int advance)
```
While I appreciate that `int` is often understood as interrupt and brevity is important, I think it would make sense to name this one `handle_interrupt`.

aktau commented on src/nvim/getchar.c:
```diff
@@ -2025,8 +2023,6 @@ static int8_t get_key_from_user(int *cp, int *timedoutp, int *mode_deletedp,
-  int new_wrow = curwin->w_wrow;
   if (       advance
       && typebuf.tb_len == 1
       && typebuf.tb_buf[typebuf.tb_off] == ESC
```
(A bit below, not your changes.)

A function that both modifies `curwin->w_w(col|row)` and is supposed to return a new version of it? I'm surprised you actually cut through the full logic of this :).

aktau commented on src/nvim/getchar.c:
```diff
@@ -2011,9 +2011,7 @@ static int look_in_typebuf(int *mapdepthp, int *keylenp,
-// Returns 0 to continue, 1 to break, -1 to carry on in the loop.
-// If it's found a key, leaves the new value of "c" in "*cp"
-static int8_t get_key_from_user(int *cp, int *timedoutp, int *mode_deletedp,
+static int ins_esc_special_case(int *new_wcolp, int *new_wrowp, int *mode_deletedp,
```
Even for static functions, I very much appreciate a comment on what the function is supposed to accomplish. For me, it's not necessary to fully comment out the variables with name and in formal setting.

```
// Does A. Puts X in new_wcolp and Y in new_wrowp. mode_deletedp is Z after ...
static int ins_esc_special_case()
```

aktau replied:
What I'm trying to say is, you made all this effort to understand this function and make it simpler, but it's still not simple enough to be obvious to the reader. It needs to be consumed in byte sized chunks, and when I see a function with a comment, I'm less tempted to actually go read it and can continue on with the current state in my head.

aktau commented on src/nvim/getchar.c:
```diff
@@ -2250,23 +2250,26 @@ static int get_key_from_user(int *timedoutp, int *mode_deletedp,
-  if (bytes_read == NUL) {                 /* no character available */
+  if (bytes_read < 0) {
+    /* end of input script reached will return -1 below */
+  } else if (bytes_read == NUL) {
```
Wait, `bytes_read` is supposed to represent a character?! That's not a very good variable name...

aktau replied:
I'd expect it to mean "number of bytes read", but if you're comparing it to `NUL`, that must not be true.

hardenedapple replied:
Thanks for the catch -- I'll fix that now. By way of explanation...

Checking for `NUL` is something left over from the original implementation that I should have changed when I split uses of a variable apart.
At this point a variable that represented a character was being reused to store the number of bytes read.
i.e. this test is checking that no bytes have been read.

aktau commented:
I've only point sampled, but otherwise LGTM. I trust that you've analyzed this function correctly, and tested it out in your local neovim instance as well.

hardenedapple committed:
Suggestions from aktau

Improvements in comments.
Rename handle_int() -> handle_interrupt()
Show keylen is set to 0 in find_typed_map() when no mapping found.
Replace NUL with 0.

aktau commented on src/nvim/getchar.c:
```diff
@@ -1847,7 +1849,7 @@ static mapblock_T *find_typed_map(const bool timedout, const int local_State,
   }
 
-  int keylen = *keylenp;
+  int keylen = 0;
```
Seems like keylen is only used much later in the code. Declaration could be moved.

aktau commented on src/nvim/getchar.c:
```diff
@@ -1847,7 +1849,7 @@ static mapblock_T *find_typed_map(const bool timedout, const int local_State,
+  int keylen = 0;
   mapblock_T *mp = NULL;
   mapblock_T *mp2 = NULL;
   int nolmaplen;
```
Can't comment below, but at the end of the function I see:

```
  // If no partly match found, use the longest full
  // match.
  if (keylen != KEYLEN_PART_MAP) {
    mp = mp_match;
    keylen = mp_match_len;
  }
  *keylenp = keylen;
  return mp;
```

```
// If no partly match found, use the longest full match.
if (keylen != KEYLEN_PART_MAP) {
  return mp_match;
}
*keylenp = keylen;
return mp;
```

While rewriting this I realized that `keylen = mp_match_len` is a completely unused statement (don't we have warnings for this?).

The fact that this exists tells me fishy stuff may be going on.

aktau replied:
Wow, scratch this. I'm stupid. My "improved" version does something different, and the initial `keylen = mp_match_len` wasn't useless. Still, I think the following would be an improvement:

```c
// If no partly match found, use the longest full match.
if (keylen != KEYLEN_PART_MAP) {
  *keylenp = mp_match_len;
  return mp_match;
}
*keylenp = keylen;
return mp;
```

aktau commented:
As a general note, if you're already touching comments in some way. I usually find it much neater to reformat comment blocks (if they don't contain structured text) so that they read better. I have a couple of mappings for this:

```
" Rewrap the current paragraph.
nnoremap <leader>q gqip
" Rewrap the current comment block. Necessary because paragraphs don't work
" properly inside of comment blocks, so we re-use tpope's comment block text
" object. nnoremap doesn't work. I have _no_ idea why, maybe because gc isn't
" builtin?. It appears to set a cursor to the end of the block, which I fix with
" marks. Overrides mark `q`, which I hope I wasn't using. Normally I'd use the
" `'` mark but that gets overriden during the `gc` movement, probably because of
" a jump.
nmap <leader>e mjgqgc`j
vmap Q gq
nmap Q gqap
```

I generally use `<leader>q` and `<leader>e`. The latter is not perfect but will leave surrounding code alone, which is very useful. `<leader>q` always does the right thing, but if code is adjacent to the comments, then the code will also be formatted. Which means you have to manually create a newline between the code and the comments. `<leader>e` is my attempt to fix that.

aktau commented on src/nvim/getchar.c:
```diff
@@ -1806,13 +1806,15 @@ static bool expand_matched_map(mapblock_T *mp, const int keylen, int *mapdepthp)
 {
   int temp_c = typebuf.tb_buf[typebuf.tb_off];
+  // Mainly to make it clear this is always 0 when calling this function.
+  assert(*keylenp == 0);
```
Perhaps it would be better to document that `*keylenp` will be set to `0` if no (partial) mapping could be found.

I only cursorily looked at the callchain of this function, but it seems like no part of the code relies on the fact that `*keylenp` is unchanged. So making it always changed (and thus always defined) makes the function easier to read.

philix commented:
Splitting functions for the sake of splitting functions doesn't necessarily make code easier to understand. I used to do that a lot (I even recommended it here in the early days of Neovim) until I realized it doesn't really help that much. 

This article from John Carmack helped me learning this.
http://number-none.com/blow/blog/programming/2014/09/26/carmack-on-inlined-code.html

The examples used to argue in favor of the "Extract Method" refactoring pattern always extract pure functions and that's nice.

I recommend function extraction in only two scenarios:

1) The extracted function is pure (the result depends solely on the paramters of the function)
2) The function will be reused -- your code calls it more than once.

> The real enemy addressed by inlining is unexpected dependency and mutation of state, which functional programming solves more directly and completely. However, if you are going to make a lot of state changes, having them all happen inline does have advantages; you should be made constantly aware of the full horror of what you are doing. When it gets to be too much to take, figure out how to factor blocks out into pure functions (and don.t let them slide back into impurity!).

philix commented on src/nvim/getchar.c:
```diff
@@ -1577,729 +1577,868 @@ vungetc ( /* unget one character (can only be done once!) */
+      }
+    }
+  }
+}
```
This function touches many globals is deeply coupled with the current state of the world. It's called only once and there is no other scenario where this function would be called and do the right thing.

philix commented on src/nvim/getchar.c:
```diff
@@ -1687,31 +1754,7 @@ static int vgetorpeek(int advance)
         keylen = 0;
         if (got_int) {
-          /* flush all input */
-          c = inchar(typebuf.tb_buf, typebuf.tb_buflen - 1, 0L,
```
`flush_all_input(typebuf)` could be a reusable function: it mutates state, but has clear semantics.

philix commented on src/nvim/getchar.c:
```diff
@@ -2312,7 +2317,7 @@ static int vgetorpeek(const int advance)
 
     /* if advance is FALSE don't loop on NULs */
-  } while (c < 0 || (advance && c == NUL));
+  } while (!exiting);
```
`exiting` is derived state from `c` and `advance`. When you introduce derived state you create the possibility of inconsistency --  e.g. `c < 0`, but `exiting` wasn't set correctly.

hardenedapple replied:
Some of the commits I have here are essentially "showing my workings", where I've split up the change I want to make into some changes that are much easier to verify are correct.

This is one of the commits in the series that splits the multiple uses of the variable `c` into separate variables.
Initially it was used to store each of

- The final character to return
- The number of bytes read from the user
- A `-1` put there simply to say "continue on in the loop"

at various points in the `while()` loop.

I found this very confusing, which is why I split it up.
The `exiting` variable is a temporary one (removed in later commits), that removes one of the uses, so that the others are easy to reason about for future changes.

philix commented on src/nvim/getchar.c:
```diff
@@ -1721,7 +1721,11 @@ static int vgetorpeek(const int advance)
+
+    if (got_int) {
+      c = handle_int(advance);
+      exiting = true;
```
Wasn't all cases of setting `exiting = true` covered in the last commit?

hardenedapple replied:
All cases were covered, but I wanted to be explicit about where the function was exiting. I thought this made it much easier when reasoning about the control flow.
Again, this is a temporary change that makes it easier to verify other commits.

philix commented on src/nvim/getchar.c:
```diff
@@ -2080,7 +2080,7 @@ static int vgetorpeek(const int advance)
             xfree(save_m_str);
             if (i == FAIL) {
-              break;
+              continue;
```
If `break` works here, `break` is preferable. Unless you found a bug.

hardenedapple replied:
The thing is that `break` breaks out of the `for(;;)` loop, and into the `while()` loop. Because of the current state in the function this immediately goes back into the `for(;;)` loop and does no different from a `continue` (except in the special case mentioned in the commit message).

Having two different lines implementing the same functionality seems unnecessary.

philix commented on src/nvim/getchar.c:
```diff
@@ -2072,16 +2069,13 @@ static int vgetorpeek(const int advance)
             }
             xfree(save_m_keys);
             xfree(save_m_str);
-            if (i == FAIL) {
```
This would check the `i` returned by `ins_typebuf`

hardenedapple replied:
Yes it would.
I made this change because after the changes of the previous commit, you can see that the actions in the case of failure and success are just to `continue`, so there's no point in checking what happened.

philix commented on src/nvim/getchar.c:
```diff
@@ -1970,7 +1970,7 @@ static int vgetorpeek(const int advance)
                 del_typebuf(1, 0);
               }
-              exiting = !(c < 0 || (advance && c == NUL));
+              exiting = true;
```
My argument against derived state holds true here. This is more error-prone and requires a non-trivial _proof of correctness_ described in your commit message.

philix commented on src/nvim/getchar.c:
```diff
@@ -1743,21 +1741,27 @@ static int vgetorpeek(const int advance)
+          }
+        }
+
+        if (typebuf.tb_len > 0) {
```
The whole idea of this commit is nice (reducing the scope of variables).

But here you removed the `else` and I can't see if it's correct or not to do so.

hardenedapple replied:
In the `if` clause we `break` out of the `for(;;)` loop (moreover `exiting` is set to `true` so that the `while()` loop will exit too).
Whether the `else` keyword is used or not, the control flow can't go through both blocks.

philix commented on src/nvim/getchar.c:
```diff
@@ -1707,620 +1706,610 @@ static int vgetorpeek(const int advance)
-    }           /* if (!character from stuffbuf) */
 
-    /* if advance is FALSE don't loop on NULs */
-  } while (!exiting);
```
This check used to be `c < 0 || (advance && c == NUL)` and `c` is mutated in this loop body. Guaranteeing `exiting` will be set correctly is really hard. You may have done it correctly here, but it's getting much harder for someone to change this code in the future.

hardenedapple replied:
The check for `exiting` is getting removed here.
It doesn't need to be set correctly any more, whether to exit or not is now being handled by the control flow of the function (i.e. whether `break` has been used in the `for(;;)` loop).

philix commented on src/nvim/getchar.c:
```diff
@@ -1730,14 +1722,17 @@ static int vgetorpeek(const int advance)
     if (typebuf.tb_no_abbr_cnt == 0)
       typebuf.tb_no_abbr_cnt = 1;             /* no abbreviations now */
   } else {
+    int timedout = FALSE;                     /* waited for more than 1 second
```
Can you have a single PR dealing with the scope reduction changes?

Mixing scope reductions and the changes you're doing here makes it hard to review and understand that your changes preserve the current behavior of the code.

philix commented on src/nvim/getchar.c:
```diff
@@ -1644,6 +1644,342 @@ static int handle_int(const int advance)
+// Keeps track of the mapdepth in the variable pointed to by "mapdepthp".
+static int look_in_typebuf(int *mapdepthp, int *keylenp,
+    int *mp_match_lenp,
+    const int timedout, const int advance, const int local_State)
```
This function is coupled with the current state of the world and is used only once. It's not pure and it's not reusable. Finding a good name for this function is impossible, hence the `look_in_typebuf` name.

The existing code is horribly complicated but keeping it inline is better than extracting it.

hardenedapple replied:
The main point of extracting this is to mark clearly which local variables are used, and which variables are modified.
I find it much easier to reason about a function if I can know for certain that in a block of code (for example) variable `timedout` is not modified, variable `mode_deleted` is not used etc. and the main problem I have with the current `vgetorpeek()` function is that it has too much local state shared over a lot of code that doesn't need it.

As a side-benefit it does make it much easier for someone new to the code to see what's going on in `vgetorpeek()`: look in the typebuffer, if nothing there, get keys from user, ad infinitum.

The coupling to the current state of the world is surprisingly small (mainly through `typebuf`), and moreover the coupling to the action of the block that I turn into `get_key_from_user()` (both through the global state, and directly) is small too.
Other than working with `typebuf`, the only global state that the two both work with is via calls to update the ui that are put after setting `'paste'` in `look_in_typebuf()` and after the `ESC` special case and the addition of pretty characters when typing an `INSERT|CMDLINE` mode map in `get_key_from_user()`.

I agree that there are downsides in putting the code elsewhere, in a perfect world I would use a hypothetical language construct where I can just say "this block can read local variables X, and Y, and can write variables A, B".
In this case, I feel the benefits of marking what variables are used and what aren't, coupled with the ease-of-reading outweigh the downsides, but that's only my opinion.

hardenedapple commented:
Just as a note:

I did at one point think of making changes like here: https://bpaste.net/show/5a1ac540168e to demonstrate more clearly what local state is read for each function (you can see that `get_key_from_user()` doesn't use `timedout` or `mode_deleted` and that `look_in_typebuf()` doesn't use `keylen`), but decided against it on the assumption that people would find the extra assurance not worth the verbosity. That assumption could easily be wrong, so I thought I'd note it here.

hardenedapple committed:
Inline do_insert_message() (suggested by philix)

As he pointed out, there's no real benefit to having it as a
separate function.

hardenedapple committed:
Use structs and enums to specify data

Rather than pass values by reference into functions that only need to
update those values, we don't give functions that information, and
make all updates in the calling function by copying members from a
returned structure.
This makes it much clearer what information a function may be affected
by, and what information is updated.

Also use an enum in a few places, to limit the information a function is
given.

justinmk commented:
I agree with @philix, in general err on the side of no changes if the result isn't a huge improvement. If it's debatable just leave it inlined. If it's a pure function with a good name then sure, extract it.

hardenedapple commented:
@justinmk do you have a specific function in mind or is that just a statement of generalities?

justinmk commented:
@hardenedapple Just a general statement.

janlazo commented:
Vim patches 8.1.1792, 8.1.1797 refactor `vgetorpeek`. Reopen this if porting those patches is insufficient.
